Add Create/Update RPCs for Auth Connectors

rosstimothy opened the pull request:
This is phase 1 of  adding optimistic locking support to auth connectors. Server side changes will land in follow up PRs to prevent breaking e.

Contributes to #30416.

espadolini commented on api/client/client.go:
```diff
@@ -1670,13 +1670,39 @@ func (c *Client) GetOIDCConnectors(ctx context.Context, withSecrets bool) ([]typ
+	_, err := c.grpc.UpsertOIDCConnectorV2(ctx, &proto.UpsertOIDCConnectorRequest{Connector: connector})
+	// TODO(tross) DELETE IN 16.0.0
+	if err != nil && trace.IsNotImplemented(err) {
+		_, err := c.grpc.UpsertOIDCConnector(ctx, connector) //nolint:staticcheck // SA1019. Kept for backward compatibility.
```
The comment makes it seem like we're only ignoring staticcheck's SA1019 even though the whole staticcheck linter is disabled (golangci-lint doesn't support the staticcheck-specific comments 😢).
```suggestion
		_, err := c.grpc.UpsertOIDCConnector(ctx, connector) //nolint:staticcheck // Deprecated RPC used for backward compatibility.
```

espadolini commented on api/client/client.go:
```diff
@@ -1735,13 +1761,39 @@ func (c *Client) GetSAMLConnectors(ctx context.Context, withSecrets bool) ([]typ
+	_, err := c.grpc.UpsertSAMLConnectorV2(ctx, &proto.UpsertSAMLConnectorRequest{Connector: samlConnector})
+	// TODO(tross) DELETE IN 16.0.0
+	if err != nil && trace.IsNotImplemented(err) {
+		_, err := c.grpc.UpsertSAMLConnector(ctx, samlConnector) ///nolint:staticcheck // SA1019. Kept for backward compatibility.
```
```suggestion
		_, err := c.grpc.UpsertSAMLConnector(ctx, samlConnector) //nolint:staticcheck // Deprecated RPC used for backward compatibility.
```

espadolini commented on api/client/client.go:
```diff
@@ -1800,13 +1852,38 @@ func (c *Client) GetGithubConnectors(ctx context.Context, withSecrets bool) ([]t
+	_, err := c.grpc.UpsertGithubConnectorV2(ctx, &proto.UpsertGithubConnectorRequest{Connector: githubConnector})
+	// TODO(tross) DELETE IN 16.0.0
+	if err != nil && trace.IsNotImplemented(err) {
+		_, err := c.grpc.UpsertGithubConnector(ctx, githubConnector) ///nolint:staticcheck // SA1019. Kept for backward compatibility testing.
```
```suggestion
		_, err := c.grpc.UpsertGithubConnector(ctx, githubConnector) //nolint:staticcheck // Deprecated RPC used for backward compatibility.
```

espadolini reviewed (approved):
What's the plan for the item-returning upsert operation?

rosstimothy committed:
Add Create/Update/UpsertV2 RPCs for Auth Connectors

Adds new RPCs which return the modified connector from write
operations. Server side and interface changes will be don in follow
up PRs to prevent breaking e. This is the first step in enforcing
optimistic locking for auth connectors.

Contributes to #30416.
